### PR TITLE
Use a BitSet for representing target types.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Use a `BitSet` rather than a `HashSet<string>` for representing target types. Since there are only a few distinct target types, we can easily represent these as a bitfield which is faster and more memory efficient than a `HashSet`.

Suggested by @chrisforbes: https://github.com/OpenRA/OpenRA/pull/14917#discussion_r175274537, https://github.com/OpenRA/OpenRA/pull/14933#discussion_r175274613.

In the default mods, there are the following numbers of distinct target types:

| Mod | Count |
| ----- | -----: |
| d2k  | 12 |
| cnc   | 16 |
| ts     | 17 |
| ra     | 26 |

Currently we use a 32-bit integer to give us a limit of 32 - should this ever become insufficient it can always be upped to a 64-bit integer for a limit of 64.

Edit: Switched to System.Numerics.BigInteger - which means no limit!

### CODE REVIEW

**Comment by pchote** [comment]
I'm pretty confident that 32 bits won't be enough for third party mods, and I won't be surprised if some are even pushing past 64.  What is the failure mode when this overflows? Does this throw an exception, or silently fail?

**Comment by pchote** [comment]
Ping @GraionDilach who should be able to provide more insight from the modding community perspective.

**Comment by RoosterDragon** [comment]
Errors, and providers a comment with suggestions for the unlucky modder who hits this problem & searches it (could always move the comment into the error message I suppose). [Thrown here](https://github.com/OpenRA/OpenRA/pull/14943/files#diff-0cc441bd228dcf0aac4ee9f48e7dbf0dR34).

**Comment by RoosterDragon** [comment]
If we're certain modders are likely to run into this a lot, we could try defining some supersized struct of sufficient size (e.g. 2 ulongs to provide 128 bits - and implement the bit fiddling needed).

Alternatively, we could add some complexity internally so it automagically falls back to an array on surpassing 64-bits which would remove the limit.

**Comment by pchote** [comment]
IMO making this 64 bit should be good enough for now and shouldn't introduce any meaningful perf overheads.

**Comment by GraionDilach** [comment]
I can't really support this.

Modders with complex mods will likely get thrown a lot on 64 I believe. Target type filtering is the most reliable and least error-prone way to filter actors for timed stat modifications and special effect/weapon immunities. Do note that RA2 just got a weapon-based mindcontrol and I also wired Black Lotus styled remote capturing to weapons for starters (because external capturing stunning the target is horribly retarded), and then chained weapons might also want spec targettype filtering to prevent them being wasted on the ore driller (something RA2 Prism did a lot). Just the sheer amount of traits using them now should have alerted the warning bells - infiltration is something classic RA already had twice (spy/thief).

In my own modding experience, relying on explicit templates/trait definitions limited to only the actors which should be affected by an effect might end up being too complicated to handle on the per-actor/per-template level if the templates can interact with each other/affect each other's traits (a simple example would be to have Temporal and EMP in the same mod, both disabling the unit, but Temporal also granting it weapon immunity for casual weapons, and this setup applied to stuff which can fly/be paradropped which involves another targettype switch, and again, this is just RA2 with EMP) and targettypes are a really good tool to be used along with global template filtering, to allow a better edge-case handling and less chance to crash out because of a messed-up cross-template interaction.

I can't tell how many target types I am using at this point though, but I'm pretty certain if I would theoretically take Mental Omega to OpenRA, it would take me a lot of headache to keep it within the 64 targettypes limit.

**Comment by pchote** [comment]
@GraionDilach Modders with complex mods also love to hate on OpenRA's poor performance.  A balance needs to be struck between capabilities and performance.

**Comment by GraionDilach** [comment]
If you are dismissing my criticism on sight, then why have you pinged me in the first place? I can understand that this PR might be able to give out a noticable perf increase due to the comment in #14933, but that alone does not justify a 64 targettype limit considering that this was never a problem before. Immunities should use targettypes as well meaningfully and not custom armors, and that leads to that a RA2/YR project wouldn't have much room to actually work with.

Target filtering was never that big of a deal of how you're acting now.

**Comment by pchote** [comment]
I was hoping/expecting that you could give some indication as to what a reasonable limit would be for modders, and not for you to simply dismiss this PR on first sight because any limit is bad.

**Comment by RoosterDragon** [comment]
`System.Numerics.BigInteger` would allow us to remove any limit - I'll find some time to see if it performs well - if so we can sidestep the "what arbitrarily low limit would still be acceptable" quandary.

**Comment by GraionDilach** [comment]
By the time I arrived, it seemed like 128 wouldn't be an option - something I'd be okay with personally. Going above 256 is definitely shouldn't be an aim, that's an amount of mess to be maintained properly, even in mind with additional future logics with the ability of employing targettype filtering as well.

**Comment by RoosterDragon** [comment]
`BigInteger` works well and means no limit! I've updated the PR to use it. This required adding a reference to `System.Numerics`.

File: Makefile [hunk 1]
```diff
@@ -41,7 +41,7 @@ SDK         ?=
 CSC         = mcs $(SDK)
 CSFLAGS     = -nologo -warn:4 -codepage:utf8 -langversion:5 -unsafe -warnaserror
 DEFINE      = TRACE
-COMMON_LIBS = System.dll System.Core.dll System.Data.dll System.Data.DataSetExtensions.dll System.Drawing.dll System.Xml.dll thirdparty/download/ICSharpCode.SharpZipLib.dll thirdparty/download/FuzzyLogicLibrary.dll thirdparty/download/MaxMind.Db.dll thirdparty/download/Eluant.dll thirdparty/download/rix0rrr.BeaconLib.dll
+COMMON_LIBS = System.dll System.Core.dll System.Data.dll System.Data.DataSetExtensions.dll System.Drawing.dll System.Numerics.dll System.Xml.dll thirdparty/download/ICSharpCode.SharpZipLib.dll thirdparty/download/FuzzyLogicLibrary.dll thirdparty/download/MaxMind.Db.dll thirdparty/download/Eluant.dll thirdparty/download/rix0rrr.BeaconLib.dll
```
Comment (RoosterDragon) [inline Makefile]: Should I add `System.Numerics` here, or should I move it specifically to OpenRA.Game which is currently the only thing using it?

Comment (pchote) [reply Makefile]: Here is probably fine.  https://github.com/OpenRA/OpenRA/blob/bleed/packaging/linux/deb/DEBIAN/control#L6 will need to be updated too.

Comment (RoosterDragon) [reply Makefile]: That already seems to include `libmono-system-numerics4.0-cil` surprisingly.

Comment (pchote) [reply Makefile]: huh, so it does.

File: OpenRA.Game/Primitives/BitSet.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using BitSetSize = System.Numerics.BigInteger;
```
Comment (pchote) [inline OpenRA.Game/Primitives/BitSet.cs]: IMO `BitSetIndex` would better reflect the way this is used.

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 3]
```diff
@@ -337,15 +337,20 @@ public interface IRenderAboveShroudWhenSelected
 		bool SpatiallyPartitionable { get; }
 	}
 
+	/// <summary>
+	/// Indicates target types as defined on <see cref="Traits.ITargetable"/> are present in a <see cref="Primitives.BitSet{T}"/>.
+	/// </summary>
+	public sealed class TargetableType { TargetableType() { } }
```
Comment (pchote) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Can we rename this to `TargetType` to match previous conventions?

**Review by pchote — DISMISSED** [review]
Code changes LGTM, just two naming nits and needs a rebase.

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 3 above)
Comment (RoosterDragon) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Conflicts with the also reasonably named `TargetType` (i.e. the enum that tells you the type of a `Target`).

Comment (pchote) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Ah right, of course. I agree that `TargetableType` is our next best option.

**Review by pchote — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/088e4a73ef6ec5c10862e33f5232d533bf7616f3...f286650870cc569848065483a3204d0ad0878fcd)

**Comment by pchote** [comment]
@RoosterDragon Do we want to port the Mobile movement types and various other `HashSet<string>` type lists to use this new plumbing?

**Comment by RoosterDragon** [comment]
It's not vital from a performance perspective, however it would hardly be unwelcome. It's a bit better for memory, and also helps correctness since `BitSet<T>` can't be compared accidentally to `BitSet<U>` - whereas `HashSet<string>` can always be compared, even if they represent different types of stuff.

**Pull request merged**